### PR TITLE
To scvi tools

### PR DESCRIPTION
**Pull request by njbernstein**
Here's the update for moving to scvi-tools. Things mostly moved over smoothly.

Notable changes:
1. Removed other doublet creation methods (doublet creation method lives in scvi-tools now)
2. Removed experimental doublets option. I think this is something the end-user can handle.
3. The overall training scheme is changed for the vae. We use a tunable learning rate which changes after validation reaches a plateau
4. Moved to the `black` code style format. 
5. Removed ability to go directly from 10x data as scvi-tools no longer directly supports it

Got increased performance on pbmcs and lower performance on kidney.
Output for performance script:
```
Mean AUROC has changed by for kidney: -0.0027609993290108603
P value for metric change AUROC in experiment kidney: 0.08545176011539873
Mean average_precision has changed by for kidney: -0.007709500685913362
P value for metric change average_precision in experiment kidney: 0.027617126859031914
WARNING!
WARNING!
WARNING!
WARNING!
WARNING!
WARNING average_precision HAS GOTTEN SIGNIFICANTLY WORSE for kidney!
Mean AUROC has changed by for pbmc: -0.008306661482196498
P value for metric change AUROC in experiment pbmc: 0.0025374340489701264
WARNING!
WARNING!
WARNING!
WARNING!
WARNING!
WARNING AUROC HAS GOTTEN SIGNIFICANTLY WORSE for pbmc!
Mean average_precision has changed by for pbmc: 0.014607552525554901
P value for metric change average_precision in experiment pbmc: 0.010120285288538741
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB average_precision HAS GOTTEN SIGNIFICANTLY BETTER for pbmc!
```

Updating of softmax scores doesnt really work anymore. Softmax scores get squished to below .1. Still looking into this

### CODE REVIEW

File: solo/solo.py [hunk 1]
```diff
@@ -178,248 +116,132 @@ def main():
...
     np.savetxt(os.path.join(args.out_dir, 'logit_scores.csv'), logit_doublet_score[:num_cells], delimiter=",")
-
     np.save(os.path.join(args.out_dir, 'logit_scores_sim.npy'), logit_doublet_score[num_cells:])
 
-
     # update threshold as a function of Solo's estimate of the number of
     # doublets
     # essentially a log odds update
     # TODO put in a function
+    # currently overshrinking softmaxes
```
Comment (njbernstein) [inline solo/solo.py:244]: @davek44 this code which modulates the softmax acores currently overshrinks the softmaxes

**Comment by njbernstein** [comment]
This code depends on https://github.com/YosefLab/scvi-tools/pull/933 being accepted

File: solo/solo.py [hunk 2]
```diff
@@ -108,63 +83,26 @@ def main():
...
-        max_cell_umi_depth = np.max(cell_umi_depth)
-        if fifth * 10 < ninetyfifth:
-            print("""WARNING YOUR DATA HAS A WIDE RANGE OF CELL DEPTHS.
-            PLEASE MANUALLY REVIEW YOUR DATA""")
-        print(f"Min cell depth: {min_cell_umi_depth}, Max cell depth: {max_cell_umi_depth}")
+        scvi_data = read_h5ad(data_path)
     else:
         msg = f'{data_path} is not a recognized format.\n'
-        msg += 'must be one of {h5ad, loom, 10x directory}'
+        msg += 'must be one of {h5ad, loom}'
```
Comment (adamgayoso) [inline solo/solo.py:101]: you can easily add readers from scanpy to support 10x loading

File: requirements.txt [hunk 3]
```diff
@@ -2,7 +2,7 @@ ConfigArgParse
 pandas
 seaborn
 tqdm
-scvi==0.6.7
+scvi-tools
```
Comment (adamgayoso) [inline requirements.txt:5]: you might as well put now `scvi-tools>=0.9.0a2`, we can release an a3 soon with SOLO too

File: solo/solo.py [hunk 4]
```diff
@@ -178,248 +116,132 @@ def main():
...
+               check_val_every_n_epoch=1,
+               early_stopping_patience=30)
+    solo.train(2000,
+               lr=learning_rate*.1,
+               train_size=.8,
+               validation_size=.1,
+               check_val_every_n_epoch=1,
+               early_stopping_patience=30,
+               callbacks=[])
+    solo.save(os.path.join(args.out_dir, 'classifier.pt'))
```
[the remark this answers is not in the document]
Comment (njbernstein) [reply solo/solo.py:191]: YOU'RE EVERYWHERE!!

But actually, really appreciate the feedback!

File: solo/solo.py (hunk 2 above)
Comment (njbernstein) [reply solo/solo.py:101]: Is it fairly recent you depend on `scanpy`? Ya adding this should be no problem then

File: requirements.txt (hunk 3 above)
Comment (njbernstein) [reply requirements.txt:5]: Will do

File: solo/solo.py (hunk 2 above)
Comment (adamgayoso) [reply solo/solo.py:101]: well, we don't explicitly rely on scanpy, it's an optional req. The readers we have are directly imported from anndata, not scanpy. That said, it seems reasonable for this repo to depend on scanpy.

File: solo/solo.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (adamgayoso) [reply solo/solo.py:191]: I'm an AI bot

File: solo/solo.py [hunk 5]
```diff
@@ -178,248 +127,121 @@ def main():
...
-    vae_params['n_batch'] = 0 if params.get(
-        'ignore_batch', False) else scvi_data.n_batches
 
     # training parameters
+    batch_key = params.get('batch_key', None)
     batch_size = params.get('batch_size', 128)
     valid_pct = params.get('valid_pct', 0.1)
     learning_rate = params.get('learning_rate', 1e-3)
-    stopping_params = {'patience': params.get('patience', 10), 'threshold': 0}
+    stopping_params = {'patience': params.get('patience', 100), 'min_delta': 0}
```
Comment (davek44) [inline solo/solo.py:136]: Does this mean 100 epochs? That seems too large

File: solo/solo.py [hunk 6]
```diff
@@ -178,248 +127,121 @@ def main():
...
-
-        full_posterior = utrainer.create_posterior(
-            utrainer.model,
-            singlet_scvi_data,
-            indices=np.arange(len(singlet_scvi_data)))
-        latent, _, _ = full_posterior.sequential(batch_size).get_latent()
-        np.save(os.path.join(args.out_dir, 'latent.npy'),
-                latent.astype('float32'))
-
+        vae.load(os.path.join(args.seed, 'vae'), use_gpu=args.gpu)
```
Comment (davek44) [inline solo/solo.py:153]: Have you tested this, too?

File: solo/solo.py [hunk 7]
```diff
@@ -178,248 +127,121 @@ def main():
...
-    # concatentate
-    classifier_data = GeneExpressionDataset()
-    classifier_data.populate_from_data(
-        X=np.vstack([scvi_data.X,
-                     in_silico_doublets]),
-        labels=np.hstack([np.ravel(scvi_data.labels),
-                          np.ones(in_silico_doublets.shape[0])]),
-        remap_attributes=False)
-
-    assert(len(np.unique(classifier_data.labels.flatten())) == 2)
```
Comment (davek44) [inline solo/solo.py:312]: Did this move into scVI?

File: solo/solo.py (hunk 1 above)
Comment (davek44) [reply solo/solo.py:244]: If you point me to an example predictions output, I can help debug

File: solo/solo.py (hunk 7 above)
Comment (njbernstein) [reply solo/solo.py:312]: Yes

File: solo/solo.py (hunk 6 above)
Comment (njbernstein) [reply solo/solo.py:153]: Not yet

File: solo/solo.py (hunk 5 above)
Comment (njbernstein) [reply solo/solo.py:136]: @davek44 I upped it because I have the patience to adjust the learning rate set to 20, and I shrink the learning rate from `1e-2` to `1e-7`. I can adjust both though to be less. A patience of 5 for adjusting the learning rate and a global patience of 30?

File: solo/solo.py (hunk 1 above)
Comment (njbernstein) [reply solo/solo.py:244]: an example output: `/home/nicholas/projects/genomics/scrnaseq/doublets/analysis/results_pbmc_1/`

File: solo/solo.py (hunk 5 above)
Comment (davek44) [reply solo/solo.py:136]: How about 10 for the learning rate and 20 to stop once and for all? Also, you typically don't decrease the learning rate quite that much. I would guess 100x decrease is about right.

File: solo/solo.py (hunk 1 above)
Comment (njbernstein) [reply solo/solo.py:244]: @davek44  did you ever take a look at this?

Comment (davek44) [reply solo/solo.py:244]: Ah no, sorry. I'll queue it

**Comment by njbernstein** [comment]
Ball park same results as before:
```
Mean AUROC has changed by for kidney: 0.004320676590979344
P value for metric change AUROC in experiment kidney: 0.027617126859031914
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB AUROC HAS GOTTEN SIGNIFICANTLY BETTER for kidney!
Mean average_precision has changed by for kidney: 0.00572280091958699
P value for metric change average_precision in experiment kidney: 0.00685541503906661
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB!
NICE JOB average_precision HAS GOTTEN SIGNIFICANTLY BETTER for kidney!
Mean AUROC has changed by for pbmc: -0.0002767180325944363
P value for metric change AUROC in experiment pbmc: 0.32403843406957306
Mean average_precision has changed by for pbmc: -0.029941475393492678
P value for metric change average_precision in experiment pbmc: 0.00405655863278289
WARNING!
WARNING!
WARNING!
WARNING!
WARNING!
WARNING average_precision HAS GOTTEN SIGNIFICANTLY WORSE for pbmc!
```

**Comment by njbernstein** [comment]
@davek44  can you take a look at this? 

Recalibration is now not done by default, but I've left the option in

**Comment by davek44** [comment]
Looks good to me. One thought-- does running solo.train twice choose a different train/validation split the second time?

**Comment by njbernstein** [comment]
@davek44  ah good point. I believe it does. Let me see what the best fix for that is

**Comment by davek44** [comment]
They added learning rate plateau helper code at some point, which we should be able to use. See https://github.com/YosefLab/scvi-tools/blob/master/scvi/train/_trainingplans.py#L186

**Comment by adamgayoso** [comment]
> Looks good to me. One thought-- does running solo.train twice choose a different train/validation split the second time?

It uses the `scvi.settings.seed` so should be the same each time.

> They added learning rate plateau helper code at some point, which we should be able to use.

It would be possible to add, though slightly more complicated

**Comment by njbernstein** [comment]
sweet, so no need to change that part. Any other issues @davek44 ?

**Comment by davek44** [comment]
Yea that's all that I see that looked questionable. Go ahead
